internal/util/http.go

```diff
@@ -0,0 +1,228 @@
+	updateInfo := &clientUpdateData{
+		Version: versionString,
+	}
+	marshaled, err := json.Marshal(updateInfo)
+	if err == nil {
+		updateInfoString := string(marshaled)
+		req.Header.Add(clientHeader, updateInfoString)
+	}
+
+	return req, err
+}
+
+// getDurationFromEnvOrConstant will either return the default duration
+// or get the duration (as string) from the environment variable
+// and parse that.
+// Durations should be in the form of "{{digit}}h" for correct parsing and reasonable behavior
+// Currently, we do not set any lower/upper bounds for these durations to allow for easy testing
+func getDurationFromEnvOrConstant(log logr.Logger, envVar string,
+	defaultDuration time.Duration) time.Duration {
+
+	durationString := os.Getenv(envVar)
+	if durationString == "" {
+		log.V(1).Info(fmt.Sprintf("no %s given, using default", envVar))
+		return defaultDuration
+	}
+	duration, err := time.ParseDuration(durationString)
+	if err == nil {
+		return duration
+	}
+	log.V(1).Info(fmt.Sprintf("error parsing %s, using default", envVar))
+	return defaultDuration
+}
+
+func getBackoff(log logr.Logger) wait.Backoff {
+	var stepsInt int
+	var err error
+
+	steps := os.Getenv(updateCheckMaxRetriesEnvVar)
+	if steps != "" {
+		stepsInt, err = strconv.Atoi(steps)
+	}
+	if steps == "" || (steps != "" && err != nil) {
+		stepsInt = updateCheckMaxRetries
+	}
+
+	return wait.Backoff{
+		Duration: getDurationFromEnvOrConstant(log, updateCheckMinimumEnvVar, updateCheckMinimum),
+		Factor:   float64(2),
+		Steps:    stepsInt,
+		Cap:      getDurationFromEnvOrConstant(log, updateCheckCapEnvVar, updateCheckCap),
+	}
+}
+
+func checkForUpdates(versionString string, backoff wait.Backoff) (string, error) {
+	var res *http.Response
+	var err error
+	var bodyBytes []byte
+
+	// Prep request
+	req, err := http.NewRequest("GET",
+		config.DefaultFromEnv(updateCheckURL, updateCheckURLEnvVar),
+		nil)
+	if err == nil {
+		req, err = addHeader(req, versionString)
+	}
+
+	// wait.ExponentialBackoff will retry the func according to the backoff object until
+	// (a) func returns done as true or
+	// (b) the backoff settings are exhausted,
+	// i.e., the process hits the cap for time or the number of steps
+	// The anonymous function here sets certain preexisting variables (res, err)
+	// which are then used by the surrounding `checkForUpdates` function as part of the return
+	if err == nil {
+		_ = wait.ExponentialBackoff(
+			backoff,
+			func() (done bool, backoffErr error) {
+				// We can't close the body of this response in this block since we use it outside
+				// so ignore this linting error
+				res, err = client.Do(req) //nolint:bodyclose
+				// This is a very basic check, ignoring nuances around
+				// certain StatusCodes that should either prevent or impact retries
+				if err == nil && res.StatusCode == http.StatusOK {
+					return true, nil
+				}
+				if err == nil {
+					err = fmt.Errorf("received StatusCode %d", res.StatusCode)
```

Comment:
I'm doing the `client.Do` inside the `wait.ExponentialBackoff`, but since ExponentialBackoff takes a [conditionFunc](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#ConditionFunc) that only returns a bool + err, I decided to assign the results of the client.Do to preexisting vars, so I could check them outside of the wait.ExponentialBackoff.

(This feels like it's verging on breaking the Golang precept "share memory by communicating, don't communicate by sharing memory," but only _feels_ like that to me.)